### ISSUE TITLE
Fix automatic HiDPI scaling on Linux

### ISSUE THREAD

**Issue by cristatus** [apache/netbeans#7780]
### Apache NetBeans version

Apache NetBeans 23 release candidate

### What happened

Netbeans 23 launcher auto-detects HiDPI and set's `GDK_SCALE=2`. However, some jdk especially 21 and JBR distributions handle scaling naively. Setting  `GDK_SCALE=2` in launch script results in 4x scaling.

### How to reproduce

Use temurin jdk 21 or Jetbrain JDK 21 and luanch netbeans 23. The netbeans UI on HiDPI screen with 2x scale looks like 4x scaled.

### Did this work correctly in an earlier version?

Apache NetBeans 22

### Operating System

Arch Linux

### JDK

21

### Apache NetBeans packaging

Apache NetBeans binary zip

### Are you willing to submit a pull request?

Yes

**Comment by mbien** [comment]
`GDK_SCALE` might be a workaround from JDK 8 times, before JDP 263. Although the JDK does check the value too:
https://github.com/openjdk/jdk/blob/ab06a878f888827026424530781f0af414a8a611/src/java.desktop/unix/native/common/awt/systemscale/systemScale.c#L41-L52

If scaling is now detected correctly on JDK 17+ the extra launcher logic should be probably removed. There are going to be some edge cases with multi screen setups where not all screens need scaling etc. If it turns out that it is still needed, we might have to set `-Dsun.java2d.uiScale.enabled=false` too whenever `GDK_SCALE` is set to prevent double scaling issues.

btw NetBeans is currently only tested on OpenJDK re-distributions, experience may vary if you run it on forks

**Comment by mbien** [comment]
@sijskes if you don't set `J2D_UISCALE` via your profile, will NB scale correctly with 2x?


> When i'm developing a desktop application, that 'exec's another GUI application, like via xdg-open, the environment of that process inherits the GDK_SCALE, and in for instance 'atril' (pdf viewer) this results in 4x scaling.

you could probably set the scaling env var to 1 if you invoke the application, no?

In any case this sounds like a separate issue to me. Going to rename this issue to make it more clear that this is about JDK forks which try to influence scaling and behave differently to OpenJDK distributions.

**Comment by sijskes** [comment]
> [@sijskes](https://github.com/sijskes) if you don't set `J2D_UISCALE` via your profile, will NB scale correctly with 2x?

yes

> > When i'm developing a desktop application, that 'exec's another GUI application, like via xdg-open, the environment of that process inherits the GDK_SCALE, and in for instance 'atril' (pdf viewer) this results in 4x scaling.
> 
> you could probably set the scaling env var to 1 if you invoke the application, no?

yes i could. or remove the variable from the environment. But if the application then runs on a platform that uses GDK_SCALE it wouldnt scale correctly.


> In any case this sounds like a separate issue to me. Going to rename this issue to make it more clear that this is about JDK forks which try to influence scaling and behave differently to OpenJDK distributions.

Ok, shall i open another?

**Comment by mbien** [comment]
>> if you don't set J2D_UISCALE via your profile, will NB scale correctly with 2x?

>yes

in that case it sounds like it could be handled by the same mechanism proposed in the above linked issue

**Comment by sijskes** [comment]
> > > if you don't set J2D_UISCALE via your profile, will NB scale correctly with 2x?
> 
> > yes
> 
> in that case it sounds like it could be handled by the same mechanism proposed in the above linked issue

No, it doesn't. Setting GDK_SCALE, causes X programs, invoked from program under development, to inherit this, and give 4x scaling. Removing GDK_SCALE is no option, because program invoked, at deployment stage, might need GDK_SCALE. 

Using J2D_UISCALE instead of GDK_SCALE limits the behaviour to java programs only.

**Comment by sijskes** [comment]
> the launcher could check both env vars

Shouldnt an IDE offer a environment very similar to deployment?  What your are implying is fixing the IDE bugs in de software developed.

**Comment by mbien** [comment]
I have nothing against J2D_UISCALE or GDK_SCALE, switching to the first _might_ be now an option since JDK requirement is 17+ but testing has to confirm it. What the launcher should likely not do is to _overwrite_  already set env vars when they were customized - this applies to both vars.

**Comment by sijskes** [comment]
> I have nothing against J2D_UISCALE or GDK_SCALE, switching to the first _might_ be now an option since JDK requirement is 17+ but testing has to confirm it. What the launcher should likely not do is to _overwrite_ already set env vars when they were customized - this applies to both vars.

I solved it by using  "export J2D_UISCALE=2" in profile script. That has the least amount of impact, and works. Moving over to #8243
